### PR TITLE
LVM-434: Implement basic user auth gating

### PR DESCRIPTION
**Pull request by zack**
https://cs5500-jira.ccs.neu.edu/browse/LVM-434

This PR does 4 things at the application level:

1. It hides links that the user can't use (like the Tutors link on the Welcome page when logged in as a tutor)
2. It _scopes_ all calls for models for which the user should only see a subset using the new `.of` method on those models. This ensures that, when this is used, it will, for example, only show a tutor the students to which they are mapped.
3. It associates users with tutors and coordinators to make the above possible.
4. It blocks out certain controller actions (and even whole controllers) based on the user's role. Look for `before_do :ensure_admin!` at the top of certain controllers.

There are a few ways this needs to be tested, and I am asking _everyone_ to review this pull request. I don't want to merge it until it's been _approved_ by everyone.

1) Look through the code and look for controller actions that need to be blocked from certain roles.
2) Look through the code and look for ActiveRecord calls (ex: `Student.all`) that need to be scoped.
3) Move through the application with different privileges and try to break stuff. Eg. try to manually go to _localhost:3000/coordinators_ when logged in as a tutor.

One problem I had when writing this was that I wasn't exactly sure who should be allowed to do what. I know that this knowledge exists in the team, it's just poorly documented. So consider the nuanced stuff that you know about, like assessments or hours or whatever.

Thanks everyone!
Remember, security is important!

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,4 +1,19 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
   before_action :authenticate_user!
+
+  def ensure_coordinator!
+    deny_access unless current_user.coordinator? || current_user.admin?
```
Comment (xingyif) [inline app/controllers/application_controller.rb]: It is more informative if we call this function `ensure_coordinator_or_admin!`. Less confusing when someone else is using it.

File: app/controllers/assessments_controller.rb [hunk 2]
```diff
@@ -3,7 +3,7 @@ class AssessmentsController < ApplicationController
 
   # GET /assessments
   def index
-    @assessments = Assessment.all
+    @assessments = Assessment.all || []
```
Comment (xingyif) [inline app/controllers/assessments_controller.rb]: From what I read online, `|| []` ensures Assessment.all is an array. Why is this necessary? We already know it is an array of Assessments.

File: db/fixtures/coordinators.rb [hunk 3]
```diff
@@ -1,4 +1,4 @@
-20.times do
+13.times do
```
Comment (xingyif) [inline db/fixtures/coordinators.rb:1]: curious why 13?

**Comment by jlally21** [comment]
Using Tutor Credentials:
- Tutors should not be able to add/edit Assessments for a Student
- Should Tutor be able to add a new Student?  You currently have the ability to, but its throwing an error: Couldn't find Student with 'id'=101 [WHERE "matches"."tutor_id" = ?]
- Should we give Tutor access to Affiliate information? (it currently does not)

Using Coordinator Credentials:
- should we give Coordinator access to Affiliate information? (it currently does not)
- should we give Coordinator access to other Coordinators information? (it currently does not)

Using Admin Credentials
- looks good

**Review by xingyif — COMMENTED** [review]
I have updated documentations for credentials based on my knowledge.
For more auth info, check
https://cs5500-jira.ccs.neu.edu/browse/LVM-434
or
https://cs5500.ccs.neu.edu/confluence/display/CS4500Sp16TEAM4/List+of+Requirements

File: db/fixtures/coordinators.rb (hunk 3 above)
Comment (zack) [reply db/fixtures/coordinators.rb:1]: I don't remember. I was just changing numbers around to get better and more meaningful association building during seeding without making _a million_ models.

File: app/controllers/assessments_controller.rb (hunk 2 above)
Comment (zack) [reply app/controllers/assessments_controller.rb]: More specifically, `|| []` returns the value after the `||` in case the value before it is falsey. 
`3 || [] => 3`
`nil || [] => []`
`false || 2 => 2`
You're right that it's unnecessary here given that `.all` will never return a falsey value.

**Comment by nabeel-** [comment]
I think most of the comments have been addressed already.

Tutors should:
- Not be able to edit or remove students. 
- Not be able to add new students
- Not be able to add new assessments

**Comment by zack** [comment]
Updated based on all comments. I'm not going to do any work on the student show page or the assessment index or show pages since those are being worked on by @vcantu and @xingyif anyway and that'll just be duplicate work. Going to write tests and finish this up.

**Review by nabeel- — APPROVED** [review]
